### PR TITLE
[PDI-18294] Limitation with Excel Writer step - appending data to an existing XLSX document does not work when "Stream XLSX data" is enabled.

### PR DESCRIPTION
**Pull request by smmribeiro**
Please, don't forget to squash.

@pentaho-lmartins @ssamora @bcostahitachivantara

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 29m 6s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl engine
```
---
:ok_hand: All tests passed!
 
Tests run: **3849**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman.pentaho.net/job/projects/job/pdi/job/PR-7823/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStep.java [hunk 1]
```diff
@@ -653,15 +655,11 @@ public void prepareNextOutputFile() throws KettleException {
...
-        XSSFWorkbook xssfWorkbook = new XSSFWorkbook( KettleVFS.getInputStream( data.file ) );
-        if ( meta.isStreamingData() && !meta.isTemplateEnabled() ) {
-          data.wb = new SXSSFWorkbook( xssfWorkbook, 100 );
-        } else {
-          //Initialize it later after writing header/template because SXSSFWorkbook can't read/rewrite existing data,
-          // only append.
-          data.wb = xssfWorkbook;
-        }
+      if ( XLSX.equalsIgnoreCase( meta.getExtension() ) ) {
+        // Ignore, by now, if it's to use streaming!
```
Comment (bcostahitachivantara) [inline engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStep.java:659]: The only problem seems related with performance, I do not know the cost of create a instance of *Workbook.

File: engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStepMeta.java [hunk 2]
```diff
@@ -845,9 +844,6 @@ public String buildFilename( VariableSpace space, int stepnr, int splitnr ) {
   @Override
   public void getFields( RowMetaInterface r, String name, RowMetaInterface[] info, StepMeta nextStep,
     VariableSpace space, Repository repository, IMetaStore metaStore ) {
-    if ( r == null ) {
```
Comment (bcostahitachivantara) [inline engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStepMeta.java:848]: What happens if you do not initialize the RowMetaInterface? You do not have a NullPointerException at getFields on Step or other action?

Comment (smmribeiro) [reply engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStepMeta.java:848]: 'r' is a local variable (well, a method parameter) and never used; as so, it was already an unnecessary instantiation.

File: engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStep.java (hunk 1 above)
Comment (smmribeiro) [reply engine/src/main/java/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStep.java:659]: It's not a performance issue, it's a functionality one.
An SXSSFWorkbook is created based on a previously created XSSFWorkbook; 
When we 'transform' an XSSFWorkbook into an SXSSFWorkbook, we loose the capability of reading existing data or getting information like the number of existing rows, last row, etc.
That's why the instantiation of the SXSSFWorkbook has been shifted to a later moment.

**Comment by ddiroma** [comment]
@ssamora Eligible to backport in next SP for 8.3 and 9.1

**Pull request merged**